lib/src/wallet/traits.rs

```diff
@@ -0,0 +1,590 @@
+/// domain. In the Orchard Domain bundles comprise Actions each of which contains
+/// both a Spend and an Output (though either or both may be dummies). Sapling transmissions,
+/// as implemented, contain a 1:1 ratio of Spends and Outputs.
+pub(crate) trait Bundle<D: DomainWalletExt<P>, P: Parameters>
+where
+    D::Recipient: Recipient,
+    D::Note: PartialEq,
+{
+    /// An expenditure of an ?external? output, such that its value is distributed among *this* transaction's outputs.
+    type Spend: Spend;
+    /// A value store that is completely emptied by transfer of its contents to another output.
+    type Output: ShieldedOutputExt<P, D>;
+    type Spends: IntoIterator<Item = Self::Spend>;
+    type Outputs: IntoIterator<Item = Self::Output>;
+    /// An extractive process that returns domain specific information from a transaction.
+    fn from_transaction(transaction: &Transaction) -> Option<&Self>;
+    fn outputs(&self) -> &Self::Outputs;
+    fn spends(&self) -> &Self::Spends;
+}
+
+impl<P: Parameters> Bundle<SaplingDomain<P>, P> for SaplingBundle<SaplingAuthorized> {
+    type Spend = SpendDescription<SaplingAuthorized>;
+    type Output = OutputDescription<GrothProofBytes>;
+    type Spends = Vec<Self::Spend>;
+    type Outputs = Vec<Self::Output>;
+    fn from_transaction(transaction: &Transaction) -> Option<&Self> {
+        transaction.sapling_bundle()
+    }
+
+    fn outputs(&self) -> &Self::Outputs {
+        &self.shielded_outputs
+    }
+
+    fn spends(&self) -> &Self::Spends {
+        &self.shielded_spends
+    }
+}
+
+impl<P: Parameters> Bundle<OrchardDomain, P> for OrchardBundle<OrchardAuthorized, Amount> {
+    type Spend = Action<Signature<SpendAuth>>;
+    type Output = Action<Signature<SpendAuth>>;
+    type Spends = NonEmpty<Self::Spend>;
+    type Outputs = NonEmpty<Self::Output>;
+
+    fn from_transaction(transaction: &Transaction) -> Option<&Self> {
+        transaction.orchard_bundle()
+    }
+
+    fn outputs(&self) -> &Self::Outputs {
+        //! In orchard each action contains an output and a spend.
+        self.actions()
+    }
+
+    fn spends(&self) -> &Self::Spends {
+        //! In orchard each action contains an output and a spend.
+        self.actions()
+    }
+}
+
+pub(crate) trait UnspentFromWalletTxns
+where
+    Self: Sized,
+{
+    fn unspent_from_wallet_txns(transactions: &WalletTxns) -> Vec<(Self, u64, TxId)>;
+}
+
+impl UnspentFromWalletTxns for SaplingNullifier {
+    fn unspent_from_wallet_txns(transactions: &WalletTxns) -> Vec<(Self, u64, TxId)> {
+        transactions.get_unspent_sapling_nullifiers()
+    }
+}
+
+impl UnspentFromWalletTxns for OrchardNullifier {
+    fn unspent_from_wallet_txns(transactions: &WalletTxns) -> Vec<(Self, u64, TxId)> {
+        transactions.get_unspent_orchard_nullifiers()
+    }
+}
+
+pub(crate) trait NoteData: Sized {
+    type Fvk: Clone;
+    type Diversifier;
+    type Note: PartialEq;
+    type Node: Hashable;
```

Comment:
This language `Node` to describe something that has a particular coordinate in a data-structure, is from the perspective of someone who's thinking of the IncrementalMerkle note commitment tree.

I am curious about the constraints on this name, and whether a more conceptually powerful might be available.

Reply:
I matched the name of the generic type in `librustzcash`, see `zcash_primitives::merkle_tree::IncrementalWitness`. A more conceptually powerful name may well be available. Note that the sapling concrete type is named `Node`, and the orchard concrete type is named `MerkleHashOrchard`